File: src/main/java/no/fintlabs/mapping/InstanceCollectionReferencesMappingService.java
package no.fintlabs.mapping;

import no.fintlabs.model.configuration.entities.InstanceCollectionReference;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.stream.IntStream;

@Service
public class InstanceCollectionReferencesMappingService {

    public List<InstanceCollectionReference> toEntity(List<String> instanceCollectionReferencesOrderedFromDto) {
        return IntStream.range(0, instanceCollectionReferencesOrderedFromDto.size())
                .mapToObj(index -> InstanceCollectionReference
                        .builder()
                        .index(index)
                        .reference(instanceCollectionReferencesOrderedFromDto.get(index))
                        .build()
                )
                .toList();
    }

    public List<String> toDto(Collection<InstanceCollectionReference> instanceCollectionReferencesOrdered) {
        return instanceCollectionReferencesOrdered.stream()
                .sorted(Comparator.comparingInt(InstanceCollectionReference::getIndex))
                .map(InstanceCollectionReference::getReference)
                .toList();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/mapping/InstanceCollectionReferencesMappingService.kt
package no.novari.flyt.configuration.mapping

import no.novari.flyt.configuration.model.configuration.entities.collection.InstanceCollectionReference
import org.springframework.stereotype.Service

@Service
class InstanceCollectionReferencesMappingService {
    fun toEntity(instanceCollectionReferencesOrderedFromDto: List<String>): List<InstanceCollectionReference> =
        instanceCollectionReferencesOrderedFromDto.mapIndexed { index, reference ->
            InstanceCollectionReference
                .builder()
                .index(index)
                .reference(reference)
                .build()
        }

    fun toDto(instanceCollectionReferencesOrdered: Collection<InstanceCollectionReference>): List<String> =
        instanceCollectionReferencesOrdered
            .sortedBy { it.index }
            .mapNotNull { it.reference }
}
